spec/tictactoe_spec.rb

```diff
@@ -50,69 +50,116 @@ def is_row_outputed(io:, row:)
 
-        expect( user_is_prompted_to_select_space ).to eql true
+        expect(user_is_prompted_to_select_space).to eql true
       end
 
       it 'adds their token to the board' do
-        fakeIO = FakeIO.new( stdin: desired_position )
+        fakeIO = FakeIO.new(stdin: desired_position)
         player_token = 'X'
         game = TicTacToe.new(io: fakeIO)
 
         updated_board = game.begin_player_turn
 
-        is_row_outputed(io: fakeIO, row: " |#{player_token}| ")
-        expect( updated_board[desired_position] ).to eql player_token
+        is_row_outputed(io: fakeIO, row:/ \|X\| /)
+        expect(updated_board[desired_position]).to eql player_token
       end
     end
 
     context 'when the user tries to place a token in an occupied space' do
       it "should prompt the user that it's taken" do
-        fakeIO = FakeIO.new( stdin: desired_position )
+        fakeIO = FakeIO.new(stdin: desired_position)
+        game = TicTacToe.new(io: fakeIO)
+
+        game.begin_player_turn
+
+        fakeIO.stdin = desired_position
+        game.begin_player_turn
+
+        user_is_prompted_position_taken = fakeIO.stdout.grep(/position.*taken/i).any?
+        expect(user_is_prompted_position_taken).to eql true
+      end
+      it 'should prompt the user to choose another position' do
+        fakeIO = FakeIO.new(stdin: desired_position)
         game = TicTacToe.new(io: fakeIO)
 
         game.begin_player_turn
 
         fakeIO.stdin = desired_position
         game.begin_player_turn
 
-        user_is_prompted_position_taken = fakeIO.stdout.grep(/[P|p]osition.*taken/).any? 
-        expect( user_is_prompted_position_taken  ).to eql true
+        user_is_prompted_to_select_space = fakeIO.stdout.grep(/select another position/i).any?
+        expect(user_is_prompted_to_select_space).to eql true
+      end
+    end
+
+    context 'when the user input is invalid' do
+      context 'if the input is outside of the specified range' do
+        it 'prompt the user to choose another position' do
+          fakeIO = FakeIO.new(stdin: -1)
+          game = TicTacToe.new(io: fakeIO)
+
+          game.begin_player_turn
+
+          user_is_prompted_position_invalid = fakeIO.stdout.grep(/invalid position/i).any?
+          user_is_prompted_to_select_space  = fakeIO.stdout.grep(/select another position/i).any?
+          expect(user_is_prompted_to_select_space).to eql true
+          expect(user_is_prompted_position_invalid).to eql true
+        end
+
+        it 'should not place the token on the board' do
+          fakeIO = FakeIO.new(stdin: -1)
+          game = TicTacToe.new(io: fakeIO)
+
+          updated_board = game.begin_player_turn
+
+          expect(updated_board.include?('X')).to_not eql true
+        end
+
+        it 'should not end the player\'s turn until a valid input is entered' do
+          fakeIO = FakeIO.new(stdin: -1)
+          game = TicTacToe.new(io: fakeIO)
+
+          player_one_token = game.current_player
```

Comment:
It looks like `current_player` was added to the public interface. What clients will use this method. Should it be public?

Reply:
It looks like it's only used for the tests. I don't think I'll need it for `Board`, `Presenter` or `MainIO`.

Reply:
I removed `#current_player` from the public interface of `TicTacToe`. All the tests relying on it have been refactored to check the current player by what the board renders.